pkg/kubelet/logs/container_log_manager.go

```diff
@@ -176,19 +176,107 @@ func (c *containerLogManager) Start() {
+				containersByName[container.Metadata.Name][i].ID = container.Id
+				containersByName[container.Metadata.Name][i].State = container.State
+				containersByName[container.Metadata.Name][i].CreatedAt = container.CreatedAt
+				break
+			}
+		}
+		if index < 0 {
+			withLog := containerWithLog{
+				ID:        container.Id,
+				State:     container.State,
+				CreatedAt: container.CreatedAt,
+				logPath:   path,
+			}
+			containersByName[container.Metadata.Name] = append(containersByName[container.Metadata.Name], withLog)
+		} else if containersByName[container.Metadata.Name][index].logPath == "" {
+			containersByName[container.Metadata.Name][index].logPath = path
+		}
+	}
+}
+
+func (c *containerLogManager) pruneDeadContainerLogs(containersByName map[string][]containerWithLog) {
+	containersToKeep := 2
+	for k, containers := range containersByName {
+		sort.Sort(containersByCreation(containers))
+		var logPath string
+		// we calculate the prefixes of logs to keep for containersToKeep containers
+		prefixesToKeep := make([]string, 0)
+		for i := range containers {
+			if i >= containersToKeep {
+				klog.V(4).Infof("not keeping log for container %s / %s", k, containers[i].ID)
+				continue
+			}
+			if containers[i].logPath != "" {
+				lastSeparator := strings.LastIndex(containers[i].logPath, string(os.PathSeparator))
+				if lastSeparator < 0 {
+					klog.Warningf("cannot find separator in %q", containers[i].logPath)
+					continue
+				}
+				runes := []rune(containers[i].logPath)
+				if logPath == "" {
+					// include the trailing separator
+					logPath = string(runes[0 : lastSeparator+1])
+				} else if containers[i].logPath != "" && !strings.HasPrefix(containers[i].logPath, logPath) {
+					klog.Warningf("%q doesn't begin with %q", containers[i].logPath, logPath)
+				}
+				prefixesToKeep = append(prefixesToKeep, string(runes[lastSeparator+1:]))
+			}
+		}
+		klog.V(4).Infof("cleaning log for %s, path %s", k, logPath)
+		pattern := fmt.Sprintf("%s*", logPath)
+		logs, err := filepath.Glob(pattern)
+		if err != nil {
+			klog.Warningf("failed to list all log files with pattern %q: %v", pattern, err)
+			continue
+		}
+		i := 0
+	outer:
+		for ; i < len(logs); i++ {
+			lastSeparator := strings.LastIndex(logs[i], string(os.PathSeparator))
+			if lastSeparator < 0 {
+				klog.Warningf("cannot find separator in %q", logs[i])
+				continue
+			}
+			runes := []rune(logs[i])
+			for _, prefix := range prefixesToKeep {
+				if strings.HasPrefix(string(runes[lastSeparator+1:]), prefix) {
+					// keep this log
+					continue outer
+				}
+			}
+			if err := os.Remove(logs[i]); err != nil && !os.IsNotExist(err) {
+				klog.Errorf("failed to remove old log %q: %v", logs[i], err)
+			} else {
+				klog.V(4).Infof("removed old log %q", logs[i])
+			}
+		}
+	}
+}
+
 func (c *containerLogManager) rotateLogs() error {
 	// TODO(#59998): Use kubelet pod cache.
 	containers, err := c.runtimeService.ListContainers(&runtimeapi.ContainerFilter{})
```

Comment:
Is it guaranteed that ListContainers will return container which logs we want to clean up, before this container is actually removed from the runtime? I mean aren't we going to have a race condition with this line: https://github.com/kubernetes/kubernetes/blob/2dc7b68b83d974deee5bcbdd913caae615fc61e4/pkg/kubelet/kuberuntime/kuberuntime_container.go#L853?

Reply:
Can you elaborate race condition in more detail ?
My understanding of race condition is between addition/modification and removal.
Say A modifies file f and B removes file f.

In the context of this bug, the logs would be removed for the second dead container. I wonder where the race is.